File: flink-streaming/src/main/java/com/example/flink/KafkaEventConsumer.java
package com.example.flink;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.serialization.AbstractDeserializationSchema;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import com.example.protocol.EventProto.Event;

import java.io.IOException;

public class KafkaEventConsumer {

    public static void main(String[] args) throws Exception {
        // Set up the streaming execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // Configuration from environment variables with defaults
        String bootstrapServers = System.getenv("KAFKA_BOOTSTRAP_SERVERS");
        if (bootstrapServers == null || bootstrapServers.isEmpty()) {
            bootstrapServers = "localhost:9092";
        }

        String topic = System.getenv("KAFKA_TOPIC");
        if (topic == null || topic.isEmpty()) {
            topic = "events";
        }

        // Configure Kafka source
        KafkaSource<Event> source = KafkaSource.<Event>builder()
                .setBootstrapServers(bootstrapServers)
                .setTopics(topic)
                .setStartingOffsets(OffsetsInitializer.earliest())
                .setValueOnlyDeserializer(new EventDeserializationSchema())
                .build();

        // Create data stream from Kafka
        DataStream<Event> events = env.fromSource(
                source,
                WatermarkStrategy.noWatermarks(),
                "Kafka Source"
        );

        // Process events
        events
                .map(event -> String.format(
                        "Event[id=%s, name=%s, timestamp=%d, data=%s]",
                        event.getId(),
                        event.getName(),
                        event.getTimestamp(),
                        event.getData()
                ))
                .print();

        // Execute the Flink job
        env.execute("Kafka Event Consumer");
    }

    /**
     * Deserialization schema for Protobuf Event messages
     */
    public static class EventDeserializationSchema extends AbstractDeserializationSchema<Event> {

        @Override
        public Event deserialize(byte[] message) throws IOException {
            try {
                return Event.parseFrom(message);
            } catch (com.google.protobuf.InvalidProtocolBufferException e) {
                // Log error and throw IOException to let Flink handle it
                System.err.println("Failed to deserialize Protobuf message: " + e.getMessage());
                throw new IOException("Invalid Protobuf message", e);
            }
        }
    }
}

File: spark-streaming/src/main/scala/com/example/spark/KafkaEventConsumer.scala
package com.example.spark

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.streaming.Trigger
import com.example.protocol.EventProto.Event
import org.slf4j.LoggerFactory

object KafkaEventConsumer {

  private val logger = LoggerFactory.getLogger(getClass)

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder()
      .appName("KafkaEventConsumer")
      .getOrCreate()

    import spark.implicits._

    // Configuration from environment variables with defaults
    val bootstrapServers = sys.env.getOrElse("KAFKA_BOOTSTRAP_SERVERS", "localhost:9092")
    val topic = sys.env.getOrElse("KAFKA_TOPIC", "events")

    // Read from Kafka
    val kafkaDF = spark.readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", bootstrapServers)
      .option("subscribe", topic)
      .option("startingOffsets", "earliest")
      .load()

    // Deserialize Protobuf messages
    val eventsDF = kafkaDF.select(
      col("key").cast("string").as("key"),
      deserializeEvent(col("value")).as("event")
    )

    // Expand the event struct
    val expandedDF = eventsDF.select(
      col("key"),
      col("event.id"),
      col("event.name"),
      col("event.timestamp"),
      col("event.data")
    )

    // Write to console for demonstration
    val query = expandedDF.writeStream
      .format("console")
      .outputMode("append")
      .trigger(Trigger.ProcessingTime("10 seconds"))
      .start()

    query.awaitTermination()
  }

  // UDF to deserialize Protobuf Event
  // Returns tuple: (id: String, name: String, timestamp: Long, data: String)
  // Returns null on deserialization failure to allow pipeline to continue processing valid messages
  val deserializeEvent = udf((bytes: Array[Byte]) => {
    if (bytes != null) {
      try {
        val event = Event.parseFrom(bytes)
        (event.getId, event.getName, event.getTimestamp, event.getData)
      } catch {
        case e: com.google.protobuf.InvalidProtocolBufferException =>
          // Log error and return null to indicate deserialization failure
          logger.error(s"Failed to deserialize Protobuf message: ${e.getMessage}", e)
          null
      }
    } else {
      null
    }
  })
}
